lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -75,33 +98,92 @@ public String toString() {
+     * If the pendingTask queue is non-empty, the head task is scheduled.
+     * There is probably no need to dump the pendingTask queue, on failure:
+     * the problem is, almost certainly, tasks that are already scheduled
+     * in the underlying executor.
+     */
     private static class ConcurrentExecutor implements CloseableExecutor {
-        private final Executor executor;
+        @NonNull
+        private final ThreadPoolExecutor executor;
+
+        @GuardedBy("this")
+        @NonNull
+        private final Queue<InstrumentedTask> pendingTasks = new LinkedList<>();
+
+        // a non-null stop latch is the flag that this executor has been stopped
+        @GuardedBy("this")
+        @Nullable
         private CountDownLatch stopLatch;
+
+        @GuardedBy("this")
         private int running;
 
-        private ConcurrentExecutor(Executor executor) { this.executor = executor; }
+        ConcurrentExecutor(@NonNull ThreadPoolExecutor executor) {
+            Preconditions.checkArgNotNull(executor, "executor");
+            this.executor = executor;
+        }
 
+        /**
+         * Schedule a task for concurrent execution.
+         * If there is insufficient room to schedule the task, safely, on the underlying
+         * executor, the task is added to the pendingTask queue and executed when spaces is available.
+         * There are absolutely no guarantees about execution order, on this executor,
+         * particularly once it fails back to using the pending task queue.
+         *
+         * @param task a task for concurrent execution.
+         * @throws ExecutorClosedException    if the executor has been stopped
+         * @throws RejectedExecutionException if the underlying executor rejects the task
+         */
         @Override
-        public synchronized void execute(@NonNull Runnable task) {
-            if (stopLatch != null) { throw new ExecutorClosedExeception("Executor has been stopped"); }
+        public void execute(@NonNull Runnable task) {
+            Preconditions.checkArgNotNull(task, "task");
 
-            try {
-                executor.execute(new InstrumentedTask(task, this::finishTask));
-                running++;
-            }
-            catch (RejectedExecutionException e) {
-                dumpServiceState(executor, e, "size: " + running);
-                throw e;
+            final int n;
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                final InstrumentedTask newTask;
+                if (spaceAvailable()) { newTask = new InstrumentedTask(task, this::finishTask); }
+                else {
+                    newTask = new InstrumentedTask(task, this::scheduleNext);
+                    pendingTasks.add(newTask);
+                }
+
+                n = pendingTasks.size();
+                if (n <= 1) { executeTask(newTask); }
```

Comment:
I'm not sure I understand everything here. Is there a case like below?
1. There are several tasks in the pendingTasks waiting to get executed.
2. There is a new task come in but at that time spaceAvailable() returns true as are some running tasks are done right before the new task was added.
3.  newTask created in line 154 will not get executed by line 161 as n > 1.